File: db4o.net/Db4objects.Db4o/Db4objects.Db4o/Internal/Marshall/ObjectReferenceContext.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

using System;
using Db4objects.Db4o.Internal;
using Db4objects.Db4o.Internal.Marshall;
using Db4objects.Db4o.Marshall;

namespace Db4objects.Db4o.Internal.Marshall
{
	public class ObjectReferenceContext : ObjectHeaderContext, IObjectIdContext
	{
		protected readonly Db4objects.Db4o.Internal.ObjectReference _reference;

		public ObjectReferenceContext(Transaction transaction, IReadBuffer buffer, ObjectHeader
			 objectHeader, Db4objects.Db4o.Internal.ObjectReference reference) : base(transaction
			, buffer, objectHeader)
		{
			_reference = reference;
		}

		public virtual int ObjectId()
		{
			return _reference.GetID();
		}

		public override Db4objects.Db4o.Internal.ClassMetadata ClassMetadata()
		{
			Db4objects.Db4o.Internal.ClassMetadata classMetadata = _reference.ClassMetadata();
			if (classMetadata == null)
			{
				throw new InvalidOperationException();
			}
			return classMetadata;
		}

		public virtual Db4objects.Db4o.Internal.ObjectReference ObjectReference()
		{
			return _reference;
		}

		protected virtual Db4objects.Db4o.Internal.ByteArrayBuffer ByteArrayBuffer()
		{
			return (Db4objects.Db4o.Internal.ByteArrayBuffer)Buffer();
		}
	}
}

File: db4oj/core/src/com/db4o/internal/marshall/ObjectReferenceContext.java
/* Copyright (C) 2008  Versant Inc.  http://www.db4o.com */

package com.db4o.internal.marshall;

/**
 * @exclude
 */
import com.db4o.internal.*;
import com.db4o.marshall.*;

public class ObjectReferenceContext extends ObjectHeaderContext implements ObjectIdContext{

    protected final ObjectReference _reference;

    public ObjectReferenceContext(Transaction transaction, ReadBuffer buffer,
        ObjectHeader objectHeader, ObjectReference reference) {
        super(transaction, buffer, objectHeader);
        _reference = reference;
    }

    public int objectId() {
        return _reference.getID();
    }

    public ClassMetadata classMetadata() {
        final ClassMetadata classMetadata = _reference.classMetadata();
        if (classMetadata == null) {
        	throw new IllegalStateException();
        }
		return classMetadata;
    }

    public ObjectReference objectReference() {
        return _reference;
    }
    
    protected ByteArrayBuffer byteArrayBuffer() {
        return (ByteArrayBuffer)buffer();
    }

}